utt/signature-processor/src/sigProcessor.cpp

```diff
@@ -0,0 +1,181 @@
+                           concordUtil::Timers& timers)
+    : msgsCommunicator_{msgsCommunicator},
+      timers_{timers},
+      repId_{repId},
+      n_{n},
+      threshold_{threshold},
+      timer_handler_timeout_{timer_handler_timeout} {
+  msgHandlers->registerMsgHandler(bftEngine::impl::MsgCode::Reserved, [&](bftEngine::impl::MessageBase* message) {
+    messages::PartialSigMsg* msg = (messages::PartialSigMsg*)message;
+    onReceivingNewPartialSig(msg->getSigId(), msg->idOfGeneratedReplica(), msg->getPartialSig());
+    delete msg;
+  });
+  timeout_handler_ = timers_.add(std::chrono::milliseconds(timer_handler_timeout_),
+                                 concordUtil::Timers::Timer::RECURRING,
+                                 [&](concordUtil::Timers::Handle h) {
+                                   (void)h;
+                                   const auto entries = jobs_.getEntries();
+                                   for (const auto& [id, entry] : entries) {
+                                     (void)id;
+                                     if (entry.start_time == 0) continue;
+                                     auto now = std::chrono::steady_clock::now().time_since_epoch().count();
+                                     if (now - entry.start_time >= entry.job_timeout_ms_) {
+                                       onJobTimeout(entry);
+                                     }
+                                   }
+                                 });
+}
+void SigProcessor::processRCMSignature(uint64_t sigId,
+                                       const libutt::api::types::CurvePoint& pidHash,
+                                       const libutt::api::Commitment& rcm1,
+                                       const libutt::api::types::CurvePoint& s2,
+                                       std::shared_ptr<libutt::api::Registrator> signer,
+                                       uint64_t job_timeout) {
+  jobs_.registerEntry(sigId, [signer, rcm1, s2, pidHash](uint64_t rid, const libutt::api::types::Signature& sig) {
+    (void)sig;
+    (void)rid;
+    return true;
+  });
+  auto partial_sig = signer->signRCM(pidHash, s2, rcm1);
+  std::vector<uint64_t> sigIds = {sigId};
+  handleSigJob(sigIds, {partial_sig.second}, job_timeout);
+}
+
+void SigProcessor::handleSigJob(std::vector<uint64_t> sigIds,
+                                const std::vector<libutt::api::types::Signature>& partial_sigs,
+                                uint64_t job_timeout) {
+  if (sigIds.size() != partial_sigs.size()) return;
+  for (size_t i = 0; i < sigIds.size(); i++) {
+    auto sigId = sigIds.at(i);
+    auto& partial_sig = partial_sigs.at(i);
+    auto& jobEntry = jobs_.get(sigId);
+    auto& validator = jobs_.getJobValidator(sigId);
+    std::vector<uint16_t> invalids;
+    for (auto& [id, psig] : jobEntry.partial_sigs) {
+      if (!validator((uint16_t)id, psig)) invalids.push_back((uint16_t)id);
```

Comment:
Why do you need here to validate the signatures as it's being done on the on message?
what is the scenario that can lead to stored invalid msgs?

Reply:
If the job leader has started the signature processing a bit later than others, he might get partial signatures that it doesn't have the data to validate.
Hence, once it has the data, it needs to validate everything it has already got

Reply:
cool ,thanks